### ISSUE TITLE
Epic: Behavioral Cloning (BC) trainer — supervised imitation learning from demonstrations

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#161]
## Background

Thrust's algorithm portfolio after the A2C epic (#150) covers:
- **PPO** (on-policy, clipped surrogate, single + multi-agent)
- **DQN** (off-policy, discrete)
- **SAC** (off-policy, continuous control)
- **A2C** (on-policy, single-update-per-rollout actor-critic)
- **PSRO / NFSP** (multi-agent meta-solvers)

**Gap**: ROADMAP Milestone 6 ("Algorithm Expansion") still has one unchecked named-classic deliverable:

> - [ ] Imitation learning (behavioral cloning)

**Behavioral Cloning (BC)** is supervised imitation learning: given a dataset of expert `(observation, action)` pairs, train a policy to reproduce the expert's actions via a plain supervised loss (cross-entropy for discrete actions). It is the simplest imitation-learning method and a clean pedagogical baseline — no environment interaction during training, no reward, no bootstrapping; just supervised classification over demonstrations.

This is `/architect` work because it makes a concrete **demo-source** decision (where expert `(obs, action)` pairs come from) and a **dataset-vs-buffer** reuse decision, then stands up a brand-new supervised training loop that differs structurally from every existing RL trainer (epoch loop over a fixed dataset, no rollout collection inside the optimizer step).

Closes the last named classic algorithm in ROADMAP Milestone 6.

### 1. Demo-collection surface — generate demos from a trained A2C policy rolling out on CartPole

**Decision: produce expert `(obs, action)` pairs in-process by rolling out an already-trained `MlpBurnPolicy` (greedy/argmax over `forward` logits) on a `CartPole` `EnvPool`, collecting `(obs, action)` into a small in-tree `Demonstrations` dataset struct.** Considered alternatives:

- **(a) Roll out a trained A2C/PPO policy on CartPole** — CHOSEN. The repo already solves CartPole with A2C (#150) and PPO. The BC example/test can train an A2C "expert" for a fixed budget (or load a quick-converged one), roll it out greedily to harvest demos, then clone. This is the lowest-friction in-tree approach: it reuses `MlpBurnPolicy`, `EnvPool`, and the exact rollout pattern already in `examples/games/cartpole/train_cartpole_a2c.rs`. It also gives a self-contained, seeded, reproducible demo source with no external files.
- **(b) Scripted/heuristic CartPole expert** — viable (CartPole has a trivial PD-style heuristic), but it introduces a hand-tuned controller that is not exercised anywhere else and would need its own correctness story. Rejected as extra surface for no reuse.
- **(c) Load demos from a file** — useful long-term but adds a serialization format + fixture files + a loader to maintain, and there is no existing demo file format in the repo. Rejected for the first cut; the `Demonstrations` struct is designed so a `from_file` constructor can be added later without changing the trainer.

**Why this is sound for BC specifically**: BC's quality ceiling is the expert's quality. An A2C policy that reaches the CartPole solved bar (mean length >= 195) is a strong expert, so a successfully cloned policy can plausibly reach a meaningful fraction of expert return (the slow eval bar in decision 3).

### 2. Demonstrations dataset — small new struct, NOT the rollout/replay buffer

**Decision: add a small `Demonstrations` dataset struct (likely `src/buffer/demonstrations.rs` or `src/train/bc/dataset.rs`), NOT a reuse of `RolloutBuffer`/`ReplayBuffer`.** Rationale:
- A BC dataset is just a flat, fixed collection of `(obs: Vec<f32> [n, obs_dim], action: Vec<i64> [n])` supervised examples. It has no rewards, no dones, no advantages/returns, no GAE, no next-state, no priorities — every field `RolloutBuffer`/`ReplayBuffer` carries is dead weight here, and forcing the BC loop through those APIs would be confusing.
- The dataset needs exactly two behaviors the RL buffers don't expose cleanly: (a) a constructor from collected host-side `(obs, action)` vectors, and (b) seeded minibatch index generation for shuffled supervised epochs. We can reuse PPO's `generate_minibatch_indices` (already public in `src/train/ppo`) for the shuffle, so the struct stays tiny (~60 LOC + tests).

API sketch:
```rust
pub struct Demonstrations { obs: Vec<f32>, actions: Vec<i64>, obs_dim: usize, len: usize }
impl Demonstrations {
    pub fn new(obs: Vec<f32>, actions: Vec<i64>, obs_dim: usize) -> Result<Self>; // validates len*obs_dim == obs.len()
    pub fn len(&self) -> usize;
    pub fn obs_dim(&self) -> usize;
    pub fn batch(&self, indices: &[usize], device: &B::Device) -> (Tensor<B,2>, Tensor<B,1,Int>); // gather a minibatch as tensors
}
```

### 3. `BcTrainer` + `BcConfig` — sibling `src/train/bc/` module, supervised cross-entropy loop

**Decision: a new sibling `src/train/bc/` module (`config.rs`, `loss.rs`, `trainer.rs`, `mod.rs`), mirroring the A2C module layout (#150).** It REUSES `MlpBurnPolicy` (the `forward(obs) -> (logits, value)` path; BC ignores the value head and trains only the policy/action head via the logits), `BurnOptimizer`, and PPO's `generate_minibatch_indices`. It DIVERGES from every RL trainer in its loop shape:

- **Loss** (`bc/loss.rs`): supervised **cross-entropy** between policy logits and expert action labels: `compute_bc_loss(logits: [batch, n_actions], actions: [batch]) -> scalar`, implemented as `-mean(log_softmax(logits).gather(action))` (equivalently negative log-likelihood of the expert action). No advantages, no value loss, no entropy bonus, no ratio. Discrete-only for the first cut (CartPole), matching the discrete `MlpBurnPolicy`.
- **Loop** (`bc/trainer.rs`): a supervised **epoch loop** over the fixed `Demonstrations` dataset. `BcTrainer::train_epoch(&Demonstrations, evaluate_fn) -> BcEpochStats` shuffles indices (seeded), iterates minibatches, computes cross-entropy, backprops, and steps the optimizer once per minibatch. `total_steps` counts gradient updates; `total_epochs` counts epochs. There is NO environment interaction inside the trainer — demos are passed in.
- **Stats**: `BcEpochStats { loss: f64, accuracy: f64 }` — mean cross-entropy loss and mean per-example action-match accuracy over the epoch (both finite host-side `f64`). Accuracy is BC-specific and the natural training-progress signal.
- **Ownership**: same `Option<P>` move-through-`step` pattern as `A2cTrainer`/`PPOTrainerBurn` (Burn's `Optimizer::step` consumes the module by value).

`BcConfig` (builder + `validate()`, mirroring `A2cConfig`/`SacConfig`):

| Knob | Default | Notes |
|---|---|---|
| `learning_rate` | `1e-3` | Adam supervised default |
| `batch_size` | `64` | supervised minibatch size |
| `epochs` | `10` | passes over the demo dataset |
| `seed` | `0` | bit-exact reproducibility (threads `MlpBurnConfig::with_seed` + seeded shuffle) |

`validate()`: `learning_rate > 0`, `batch_size > 0`, `epochs > 0`.

### 4. Benchmark / eval — fast always-on smoke + slow gated convergence

Mirrors the A2C two-tier convention (`tests/test_a2c_cartpole.rs`, `tests/test_sac_pendulum.rs`):

- **Always-on fast unit test** (`bc/trainer.rs` `#[cfg(test)]`): construct `BcTrainer` on `Autodiff<NdArray<f32>>`, build a tiny synthetic `Demonstrations` (a handful of `(obs, action)` rows), run ONE supervised step (or one `train_epoch` over a one-minibatch dataset), assert (a) at least one gradient step fired, (b) `loss` and `accuracy` are finite. Fast, deterministic, runs on every PR.
- **Slow convergence/eval test** (`tests/test_bc_cartpole.rs`, gated `training`, `#[ignore]`): (1) train an A2C expert on CartPole to the solved bar (or a fixed strong budget), (2) roll it out greedily to harvest demos, (3) clone with `BcTrainer`, (4) roll out the cloned policy and assert it reaches a meaningful fraction of expert return, e.g. **mean episode length >= 150 over N episodes** (random baseline ~22; expert ~195+; 150 cleanly separates "cloned the expert" from "did not learn"). Marked `#[ignore]`, opt-in via `--ignored`, exactly like the A2C bar.

### 5. Example — `train_bc_cartpole`

`examples/games/cartpole/train_bc_cartpole.rs` (registered in `Cargo.toml` like `train_cartpole_a2c`): train/obtain an A2C expert, harvest demos via greedy rollout, clone with `BcTrainer`, then report **cloned-vs-expert mean episode reward** over evaluation rollouts. Seeded end-to-end. Mirrors the A2C example's structure and feature gating (`--features training`, optional `wgpu`).

### 6. Decomposition into builder-tractable PRs

Three PRs, ordered with dependencies and bounded blast radius (all new code under `src/train/bc/` + one new dataset file + one test + one example; zero edits to existing trainers):

- **PR A — `Demonstrations` dataset + `BcConfig`** (no deps; immediately buildable). New `Demonstrations` struct (constructor + validation + `batch()` gather + unit tests) and `BcConfig` (builder + `validate()` + tests). Blast radius: new files only.
- **PR B — `BcConfig` consumer: `bc/loss.rs` + `BcTrainer` + smoke test** (depends on PR A). Cross-entropy loss + supervised epoch loop + `BcEpochStats` + always-on fast smoke test. Blast radius: new `src/train/bc/` files + a `pub mod bc;` line in `src/train/mod.rs`.
- **PR C — `train_bc_cartpole` example + slow CartPole eval test** (depends on PR B). End-to-end demo-harvest -> clone -> cloned-vs-expert report example, plus the `#[ignore]`d convergence/eval test. Blast radius: new example file + `Cargo.toml` `[[example]]` entry + new test file.

DAG: `A -> B -> C` (linear chain).

## Affected files (whole epic)
- NEW `src/train/bc/mod.rs`, `src/train/bc/config.rs`, `src/train/bc/loss.rs`, `src/train/bc/trainer.rs`
- NEW `src/buffer/demonstrations.rs` (or `src/train/bc/dataset.rs` — builder's call, default to the `bc/dataset.rs` co-location)
- NEW `tests/test_bc_cartpole.rs`
- NEW `examples/games/cartpole/train_bc_cartpole.rs`
- EDIT `src/train/mod.rs` (add `pub mod bc;` + re-exports)
- EDIT `src/buffer/mod.rs` (only if the dataset lands under `src/buffer/`)
- EDIT `Cargo.toml` (add the `[[example]]` block)
- EDIT `ROADMAP.md` (check off "Imitation learning (behavioral cloning)")

## Acceptance (epic)
- ROADMAP Milestone 6 "Imitation learning (behavioral cloning)" is checked off.
- `BcTrainer` + `BcConfig` exist as a sibling `src/train/bc/` module following the A2C pattern.
- A cloned CartPole policy reaches mean episode length >= 150 over the eval rollouts in the `#[ignore]`d test.
- `train_bc_cartpole` example runs end-to-end and reports cloned-vs-expert reward.
- Zero changes to existing trainer loops (PPO/DQN/SAC/A2C).

**Comment by rjwalters** [comment]
## Child issues + DAG

Linear chain `A -> B -> C` (each PR is bounded to new files + minimal wiring; zero edits to existing trainer loops):

| PR | Issue | Title | Depends on | Label | Buildable now |
|---|---|---|---|---|---|
| A | #164 | Demonstrations dataset + BcConfig | — | `loom:issue` | YES |
| B | #167 | cross-entropy loss + BcTrainer epoch loop + smoke test | #164 | `loom:blocked` | after #164 |
| C | #169 | train_bc_cartpole example + slow cloned-vs-expert eval test | #167 | `loom:blocked` | after #167 |

```
#164 (A: dataset + config)  --ready-->
   └─> #167 (B: loss + trainer + smoke)
          └─> #169 (C: example + slow eval; closes ROADMAP M6 imitation-learning)
```

**Immediately buildable**: #164 only.

## Key decisions (recap)
1. **Demo source**: greedy rollout of a trained A2C `MlpBurnPolicy` on CartPole `EnvPool` (lowest-friction, in-tree, seeded) — over a scripted heuristic or file loader.
2. **Dataset**: small new `Demonstrations` struct (not `RolloutBuffer`/`ReplayBuffer`); reuses PPO's `generate_minibatch_indices` for seeded shuffling.
3. **Trainer**: sibling `src/train/bc/` module (A2C layout) — supervised cross-entropy epoch loop over a fixed dataset, reuses `MlpBurnPolicy` (logits head only) + `BurnOptimizer`.
4. **Eval**: fast always-on smoke (one epoch, finite loss/accuracy) in #167; slow `#[ignore]`d test in #169 asserts cloned mean episode length >= 150.
5. **Closes** ROADMAP Milestone 6's last named-classic item (behavioral cloning).

**Comment by rjwalters** [comment]
## Behavioral Cloning epic complete ✅ — closes ROADMAP Milestone 6
- #164 (PR #170): `Demonstrations` dataset + `BcConfig`
- #167 (PR #174): cross-entropy loss + `BcTrainer` + smoke test
- #169 (PR #177): `train_bc_cartpole` example + slow eval test

End-to-end BC pipeline (train A2C expert → greedily harvest demos → clone fresh policy → eval): cloned policy reaches **~170 mean reward vs expert ~173** on CartPole (random baseline ~22), ~99% action-match. BC is Thrust's 5th single-agent algorithm; ROADMAP Milestone 6 (PPO, DQN, SAC, A2C, BC) is now complete. Closing.
